### PR TITLE
Add CRTP pattern Ref: #2342

### PR DESCRIPTION
**Pull request by AddeusExMachina**
Add crtp directory containing source code and documentation for the curiously recurring template pattern

### CODE REVIEW

File: crtp/README.md [hunk 1]
```diff
@@ -0,0 +1,134 @@
+---
+title: Curiously Recurring Template Pattern
+language: en
+---
```
Comment (iluwatar) [inline crtp/README.md:4]: check out the required yaml frontmatter at https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute

File: crtp/src/main/java/crtp/MmaFighter.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+package crtp;
+
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * MmaFighter class.
+ *
+ * @param <T> MmaFighter derived class that uses itself as type parameter.
+ */
+@Slf4j
```
Comment (iluwatar) [inline crtp/src/main/java/crtp/MmaFighter.java:34]: Check out for what else you can use Lombok for e.g. generate constructor, toString etc.

File: crtp/src/test/java/crtp/AppTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+/**
+ * Application test
+ */
+class AppTest {
+
+  @Test
+  void shouldExecuteApplicationWithoutException() {
+    assertDoesNotThrow(() -> App.main(new String[]{}));
+  }
```
Comment (iluwatar) [inline crtp/src/test/java/crtp/AppTest.java:39]: This is the only test? I think it's not sufficient.

**Review by iluwatar — CHANGES_REQUESTED** [review]

File: crtp/src/main/java/crtp/MmaFighter.java (hunk 2 above)
Comment (AddeusExMachina) [reply crtp/src/main/java/crtp/MmaFighter.java:35]: I replaced the constructor and toString() in MmaFighter with @Data annotation

File: crtp/src/test/java/crtp/AppTest.java (hunk 3 above)
Comment (AddeusExMachina) [reply crtp/src/test/java/crtp/AppTest.java:39]: There's a new test in FightTest.java file

**Review by iluwatar — APPROVED** [review]

**Comment by iluwatar** [comment]
Looks good! Thank you for the contribution 🎉

@all-contributors please add @AddeusExMachina for code

**Pull request merged**